### PR TITLE
feat: add gamma uniswap vaults

### PR DESCRIPTION
**Pull request by Crypto-One-dev**

### CODE REVIEW

**Comment by Crypto-One-dev** [comment]
Gamma Uniswap Vaults Add

WMATIC-USDC Vault
test block: 59558330

```
Mainnet Uniswap-Gamma WMATIC-USDC
Impersonating...
0xF066789028fE31D4f53B69B81b328B8218Cc0641
0x365A834103Cf143fAb443BC3db987D9cF8e461b2
0xddDF49f09928967294eca61C6593327518Bd343F
0x6a74649aCFD7822ae8Fb78463a9f2192752E5Aa2
Fetching Underlying at:  0xE583b04b9a8F576aa7F17ECc6eB662499B5A8793
New Vault Deployed:  0xa5dffa7171bd9Afb8384D3dc7C284a3b10a4Cb2A
Strategy Deployed:  0x1eC9d9A53c352B5e03d12775b1785f46421AB3f7
    Happy path
Farmer old balance: 1196371606
Farmer fToken balance: 1196371606
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
instant APR: 0 %
instant APY: 0 %
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000298742462799639
growth:  1.0002987424627996
instant APR: 52.33967948248903 %
instant APY: 68.7118235154264 %
loop  2
old shareprice:  1000298742462799639
new shareprice:  1000597754072742512
growth:  1.000298922309156
instant APR: 52.37118856414327 %
instant APY: 68.76491527100642 %
loop  3
old shareprice:  1000597754072742512
new shareprice:  1000896843417729858
growth:  1.0002989106699172
instant APR: 52.369149369486934 %
instant APY: 68.76147879157676 %
loop  4
old shareprice:  1000896843417729858
new shareprice:  1001196021363950692
growth:  1.0002989098708706
instant APR: 52.36900937652145 %
instant APY: 68.76124287602478 %
loop  5
old shareprice:  1001196021363950692
new shareprice:  1001495287911405012
growth:  1.0002989090458496
instant APR: 52.36886483285268 %
instant APY: 68.76099929199759 %
loop  6
old shareprice:  1001495287911405012
new shareprice:  1001794657269724604
growth:  1.0002989223832932
instant APR: 52.37120155297035 %
instant APY: 68.76493716019456 %
loop  7
old shareprice:  1001794657269724604
new shareprice:  1002094101855506590
growth:  1.000298908148101
instant APR: 52.36870754730809 %
instant APY: 68.76073423584157 %
loop  8
old shareprice:  1002094101855506590
new shareprice:  1002393649252153849
growth:  1.000298921424737
instant APR: 52.371033613937534 %
instant APY: 68.76465414429913 %
loop  9
old shareprice:  1002393649252153849
new shareprice:  1002693273547984889
growth:  1.0002989088129743
instant APR: 52.368824033104566 %
instant APY: 68.76093053659122 %
earned!
Overall APR: 52.43732753717587 %
Overall APY: 68.87641144122102 %
      ✔ Farmer should earn money (34999ms)
    Basic strategy functionality checks
      ✔ Deposit to vault
      ✔ withdrawAllToVault() (88ms)
      ✔ withdrawToVault(uint256 _amount) (54ms)
      ✔ unsalvagableTokens(address _token)
      ✔ Remove all deposited LP tokens (53ms)


  6 passing (45s)
```


USDC-WETH vault
test block: 59826400

```
Mainnet Uniswap-Gamma USDC-WETH
Impersonating...
0xF066789028fE31D4f53B69B81b328B8218Cc0641
0xF4489f838be09C57A2a1D22252675b38CA1daD24
0x6a74649aCFD7822ae8Fb78463a9f2192752E5Aa2
Fetching Underlying at:  0x1Fd452156b12FB5D74680C5Ff166303E6dd12A78
New Vault Deployed:  0xa5dffa7171bd9Afb8384D3dc7C284a3b10a4Cb2A
Strategy Deployed:  0x1eC9d9A53c352B5e03d12775b1785f46421AB3f7
    Happy path
Farmer old balance: 97969713619389320
Farmer fToken balance: 97969713619389320
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
instant APR: 0 %
instant APY: 0 %
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000248628490916066
growth:  1.000248628490916
instant APR: 43.559711608483724 %
instant APY: 54.548441473104845 %
loop  2
old shareprice:  1000248628490916066
new shareprice:  1000497735920803939
growth:  1.0002490455100785
instant APR: 43.632773365759896 %
instant APY: 54.661263733344725 %
loop  3
old shareprice:  1000497735920803939
new shareprice:  1000746428737040456
growth:  1.0002485690944694
instant APR: 43.54930535103065 %
instant APY: 54.532378767385566 %
loop  4
old shareprice:  1000746428737040456
new shareprice:  1000995184373545775
growth:  1.000248570096642
instant APR: 43.54948093166441 %
instant APY: 54.532649773217216 %
loop  5
old shareprice:  1000995184373545775
new shareprice:  1001244002830319926
growth:  1.000248571082717
instant APR: 43.54965369200432 %
instant APY: 54.532916426416776 %
loop  6
old shareprice:  1001244002830319926
new shareprice:  1001492871543312363
growth:  1.0002485595042656
instant APR: 43.54762514733501 %
instant APY: 54.52978542457805 %
loop  7
old shareprice:  1001492871543312363
new shareprice:  1001741815640624189
growth:  1.000248573009739
instant APR: 43.54999130627579 %
instant APY: 54.53343753072548 %
loop  8
old shareprice:  1001741815640624189
new shareprice:  1001990822558204878
growth:  1.000248573947591
instant APR: 43.55015561794939 %
instant APY: 54.533691144810994 %
loop  9
old shareprice:  1001990822558204878
new shareprice:  1002239886014029157
growth:  1.0002485685998483
instant APR: 43.54921869342955 %
instant APY: 54.53224501297762 %
earned!
Overall APR: 43.607385261864984 %
Overall APY: 54.62205001221374 %
      ✔ Farmer should earn money (100595ms)
    Basic strategy functionality checks
      ✔ Deposit to vault
      ✔ withdrawAllToVault() (404ms)
      ✔ withdrawToVault(uint256 _amount) (49ms)
      ✔ unsalvagableTokens(address _token)
      ✔ Remove all deposited LP tokens (61ms)


  6 passing (2m)
```


USDC-USDT vault
test block: 59760970

```
Mainnet Uniswap-Gamma USDC-USDT
Impersonating...
0xF066789028fE31D4f53B69B81b328B8218Cc0641
0x5BE20371FBca17Dc86BCE43d3317F72FeAd0281c
0x6a74649aCFD7822ae8Fb78463a9f2192752E5Aa2
Fetching Underlying at:  0x8c6FE430cf06e56BE7c092AD3A249BF0BcB388B9
New Vault Deployed:  0xa5dffa7171bd9Afb8384D3dc7C284a3b10a4Cb2A
Strategy Deployed:  0x1eC9d9A53c352B5e03d12775b1785f46421AB3f7
    Happy path
Farmer old balance: 8513390875
Farmer fToken balance: 8513390875
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
instant APR: 0 %
instant APY: 0 %
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000091970991523398
growth:  1.0000919709915235
instant APR: 16.113317714912156 %
instant APY: 17.4799647893185 %
loop  2
old shareprice:  1000091970991523398
new shareprice:  1000184143782779150
growth:  1.0000921643148122
instant APR: 16.14718795509429 %
instant APY: 17.519744692231566 %
loop  3
old shareprice:  1000184143782779150
new shareprice:  1000276124523649338
growth:  1.000091963806307
instant APR: 16.112058864995046 %
instant APY: 17.4784865547444 %
loop  4
old shareprice:  1000276124523649338
new shareprice:  1000368120886966792
growth:  1.0000919709678777
instant APR: 16.113313572176402 %
instant APY: 17.47995992457978 %
loop  5
old shareprice:  1000368120886966792
new shareprice:  1000460113608961951
growth:  1.0000919588700143
instant APR: 16.11119402651262 %
instant APY: 17.47747100825048 %
loop  6
old shareprice:  1000460113608961951
new shareprice:  1000552114788221796
growth:  1.0000919588677335
instant APR: 16.11119362690907 %
instant APY: 17.477470539009655 %
loop  7
old shareprice:  1000552114788221796
new shareprice:  1000644124542208335
growth:  1.0000919589820727
instant APR: 16.111213659137743 %
instant APY: 17.47749406199015 %
loop  8
old shareprice:  1000644124542208335
new shareprice:  1000736142635997551
growth:  1.0000919588608301
instant APR: 16.111192417439213 %
instant APY: 17.47746911878196 %
loop  9
old shareprice:  1000736142635997551
new shareprice:  1000828168012431356
growth:  1.000091957682463
instant APR: 16.110985967508995 %
instant APY: 17.47722669388736 %
earned!
Overall APR: 16.121874606162123 %
Overall APY: 17.490013414811244 %
      ✔ Farmer should earn money (128440ms)
    Basic strategy functionality checks
      ✔ Deposit to vault
      ✔ withdrawAllToVault() (370ms)
      ✔ withdrawToVault(uint256 _amount) (54ms)
      ✔ unsalvagableTokens(address _token)
      ✔ Remove all deposited LP tokens (49ms)


  6 passing (3m)
```

USDC-DAI vault
test block: 

```
Mainnet Uniswap-Gamma USDC-DAI
Impersonating...
0xF066789028fE31D4f53B69B81b328B8218Cc0641
0x44034e2261eDE365B20Eb18B109Ca3698C19ddc9
0x6a74649aCFD7822ae8Fb78463a9f2192752E5Aa2
Fetching Underlying at:  0x831231E16D95Eb3D54Bf2C80968F35A5F4483447
New Vault Deployed:  0xa5dffa7171bd9Afb8384D3dc7C284a3b10a4Cb2A
Strategy Deployed:  0x1eC9d9A53c352B5e03d12775b1785f46421AB3f7
    Happy path
Farmer old balance: 4401739579854105001577
Farmer fToken balance: 4401739579854105001577
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
instant APR: 0 %
instant APY: 0 %
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000093487076400903
growth:  1.0000934870764009
instant APR: 16.378935785435722 %
instant APY: 17.79228840043132 %
loop  2
old shareprice:  1000093487076400903
new shareprice:  1000187178803704143
growth:  1.0000936829691565
instant APR: 16.41325619621199 %
instant APY: 17.832703977669517 %
loop  3
old shareprice:  1000187178803704143
new shareprice:  1000280655946679998
growth:  1.0000934596493103
instant APR: 16.374130559162836 %
instant APY: 17.786630888738774 %
loop  4
old shareprice:  1000280655946679998
new shareprice:  1000374131867146558
growth:  1.0000934496932545
instant APR: 16.372386258194105 %
instant APY: 17.78457727452294 %
loop  5
old shareprice:  1000374131867146558
new shareprice:  1000467606391068133
growth:  1.0000934395652026
instant APR: 16.370611823492 %
instant APY: 17.782488219655246 %
loop  6
old shareprice:  1000467606391068133
new shareprice:  1000561086130671850
growth:  1.0000934360483105
instant APR: 16.369995663994175 %
instant APY: 17.78176281924719 %
loop  7
old shareprice:  1000561086130671850
new shareprice:  1000654568127128018
growth:  1.0000934295744177
instant APR: 16.368861437983462 %
instant APY: 17.780427514282437 %
loop  8
old shareprice:  1000654568127128018
new shareprice:  1000748048899981099
growth:  1.000093419623345
instant APR: 16.36711801005823 %
instant APY: 17.778375035725745 %
loop  9
old shareprice:  1000748048899981099
new shareprice:  1000841535408978924
growth:  1.0000934166287916
instant APR: 16.36659336428607 %
instant APY: 17.777757395035245 %
earned!
Overall APR: 16.381629618533005 %
Overall APY: 17.79546014740372 %
      ✔ Farmer should earn money (132544ms)
    Basic strategy functionality checks
      ✔ Deposit to vault
      ✔ withdrawAllToVault() (355ms)
      ✔ withdrawToVault(uint256 _amount) (51ms)
      ✔ unsalvagableTokens(address _token)
      ✔ Remove all deposited LP tokens (47ms)


  6 passing (3m)
```

File: contracts/strategies/uniswap-gamma/UniswapGammaStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+import "@openzeppelin/contracts/math/Math.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import "../../base/upgradability/BaseUpgradeableStrategyV2.sol";
+import "../../base/interface/IUniversalLiquidator.sol";
+import "../../base/interface/gamma/IHypervisor.sol";
+import "../../base/interface/gamma/IUniProxy.sol";
+import "../../base/interface/gamma/IStakingRewards.sol";
+import "../../base/interface/quickswap/IDragonLair.sol";
```
Comment (CryptJS13) [inline contracts/strategies/uniswap-gamma/UniswapGammaStrategy.sol]: Can be removed

File: contracts/strategies/uniswap-gamma/UniswapGammaStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,313 @@
+//SPDX-License-Identifier: Unlicense
+
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts/math/Math.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
```
Comment (CryptJS13) [inline contracts/strategies/uniswap-gamma/UniswapGammaStrategy.sol]: Can you copy the contract into `base/interface/uniswap` and remove the dependency?

File: package.json [hunk 3]
```diff
@@ -50,6 +50,7 @@
   },
   "dependencies": {
     "@nomiclabs/hardhat-truffle5": "^2.0.0",
+    "@uniswap/v3-core": "^1.0.1",
```
Comment (CryptJS13) [inline package.json]: As said before, please copy the contract into our own folder and remove the dependency.

**Review by CryptJS13 — COMMENTED** [review]
Few comments

**Pull request merged**